weave/trace/vals.py

```diff
@@ -350,29 +424,44 @@ def __init__(
         super().__init__(*args, **kwargs)
 
     def __getitem__(self, key: str) -> Any:
-        new_ref = self.ref.with_key(key)
-        return make_trace_obj(super().__getitem__(key), new_ref, self.server, self.root)
+        new_ref = self.ref.with_key(key) if self.ref else None
+        v = super().__getitem__(key)
+        return make_trace_obj(v, new_ref, self.server, self.root)
 
     def get(self, key: str, default: Any = None) -> Any:
-        new_ref = self.ref.with_key(key)
-        return make_trace_obj(
-            super().get(key, default), new_ref, self.server, self.root
-        )
+        new_ref = self.ref.with_key(key) if self.ref else None
+        v = super().get(key, default)
+        return make_trace_obj(v, new_ref, self.server, self.root)
 
     def __setitem__(self, key: str, value: Any) -> None:
+        full_path: tuple[str, ...]
         if not isinstance(self.ref, ObjectRef):
-            raise ValueError("Can only set items on object refs")
+            # We used to raise ValueError here to only set attributes on
+            # object refs but I don't understand why that's necessary.
```

Comment:
Anyone know why we used to raise here?

Reply:
What are the other Ref types you are now handling?

Reply:
The intent of this section is `if self.ref is None`, but I kept this comment and structure because I wasn't sure why we needed to raise.